File: services/config/profileCli.ts
/**
 * Mossen multi-profile CLI flag handler (S1-09c, D-S09-2=Z).
 *
 * 对外 (用户友好):
 *   mossen --list-model-profiles
 *   mossen --get-model-profile [<name>]              # 无 name = 当前 active
 *   mossen --set-model-profile <name>                # 切换 active profile
 *   mossen --add-model-profile <name>
 *           --provider openai-compatible
 *           --baseURL <url>
 *           --model <id>
 *           --apiKey <key>
 *           [--name <display-name>]
 *           [--scope user|project]                   # default user
 *   mossen --update-model-profile <name>
 *           [--baseURL <url>] [--model <id>]
 *           [--apiKey <key>] [--name <display-name>]
 *           [--scope user|project]
 *   mossen --set-model-profile-key <name> <key> [--scope ...]
 *   mossen --delete-model-profile <name> [--scope ...]
 *
 * 内部 (D-S09-2=Z): 全部走 services/config/profiles.ts (已建在 facade chain 上).
 *
 * 输出 contract:
 *   - stdout = JSON (UI 可解析)
 *   - 任何 dump apiKey 的字段必须用 maskApiKey 脱敏
 *   - exit 0 = 成功; exit 1 = 校验失败 / 未找到; exit 2 = 内部异常
 */

import {
  desensitizeProfile,
  desensitizeProfiles,
  getProfiles,
  getActiveProfileName,
  getProfileByName,
  setProfile,
  setActiveProfile,
  deleteProfile,
  testProfile,
  getFallbackProfile,
  listAllProfiles,
  getCurrentProfile,
  getDefaultProfile,
  migrateFallbackProfile,
  type ProfileSchema,
  type ProfileProvider,
  PROFILE_PROVIDER_VALUES,
} from './profiles.js'

const MODEL_PROFILE_FLAGS = [
  '--list-model-profiles',
  '--get-model-profile',
  '--set-model-profile',
  '--add-model-profile',
  '--update-model-profile',
  '--set-model-profile-key',
  '--delete-model-profile',
  '--test-model-profile',
  '--migrate-fallback-profile',
] as const

export type ModelProfileFlag = (typeof MODEL_PROFILE_FLAGS)[number]

export function isModelProfileFlagPresent(args: readonly string[]): boolean {
  return args.some(arg => (MODEL_PROFILE_FLAGS as readonly string[]).includes(arg))
}

function findFlag(args: readonly string[], flag: string): number {
  return args.indexOf(flag)
}

function getOptionValue(args: readonly string[], flag: string): string | undefined {
  const idx = args.indexOf(flag)
  if (idx === -1) return undefined
  const next = args[idx + 1]
  if (next === undefined || next.startsWith('--')) return undefined
  return next
}

function parseScope(args: readonly string[]): 'user' | 'project' {
  const v = getOptionValue(args, '--scope')
  if (v === 'project') return 'project'
  return 'user'
}

function emitJson(payload: unknown): void {
  process.stdout.write(`${JSON.stringify(payload, null, 2)}\n`)
}

function emitError(msg: string): void {
  process.stderr.write(`error: ${msg}\n`)
}

export async function handleModelProfileCliFlag(
  args: readonly string[],
): Promise<{ handled: boolean; exitCode: number }> {
  if (!isModelProfileFlagPresent(args)) {
    return { handled: false, exitCode: 0 }
  }

  try {
    // --list-model-profiles
    if (findFlag(args, '--list-model-profiles') !== -1) {
      const settingsProfiles = getProfiles()
      const desensitizedSettings = desensitizeProfiles(settingsProfiles)
      const all = listAllProfiles()
      const fallback = getFallbackProfile()
      const current = getCurrentProfile()
      const defaultP = getDefaultProfile()
      const allWithSource = all.map(item => ({
        name: item.name,
        source: item.source,
        profile: desensitizeProfile(item.profile),
      }))
      emitJson({
        // 兼容旧 contract: settings-持久化 profile map (apiKey 脱敏)
        profiles: desensitizedSettings,
        // 旧字段: settings 中 active 字段值 (run-time override 后是 session 名;
        // 若 active 指向不存在的 profile 或为 null → null)
        activeProfile: getActiveProfileName(),
        // S1-09 闭环新字段:
        // 完整可见 profile 列表 (含 fallback 虚拟 profile, source 区分)
        allProfiles: allWithSource,
        // fallback (env-based 虚拟 profile, apiKey 脱敏); 不存在则 null
        fallbackProfile: fallback
          ? { name: fallback.name, source: fallback.source, profile: desensitizeProfile(fallback.profile) }
          : null,
        // 当前会话实际在用的 profile (session > user > fallback)
        currentProfile: current
          ? { name: current.name, source: current.source, profile: desensitizeProfile(current.profile) }
          : null,
        // 全局默认 profile (settings active > fallback)
        defaultProfile: defaultP
          ? { name: defaultP.name, source: defaultP.source, profile: desensitizeProfile(defaultP.profile) }
          : null,
        count: Object.keys(desensitizedSettings).length,
        countAll: allWithSource.length,
      })
      return { handled: true, exitCode: 0 }
    }

    // --get-model-profile [<name>]
    {
      const idx = findFlag(args, '--get-model-profile')
      if (idx !== -1) {
        const explicitName = args[idx + 1]
        const targetName = explicitName && !explicitName.startsWith('--') ? explicitName : null
        if (targetName) {
          const p = getProfileByName(targetName)
          if (p) {
            emitJson({ name: targetName, source: 'settings', profile: desensitizeProfile(p) })
            return { handled: true, exitCode: 0 }
          }
          const fb = getFallbackProfile()
          if (fb && fb.name === targetName) {
            emitJson({ name: targetName, source: fb.source, profile: desensitizeProfile(fb.profile) })
            return { handled: true, exitCode: 0 }
          }
          emitError(`profile "${targetName}" not found`)
          return { handled: true, exitCode: 1 }
        }
        // 无 name → 当前会话实际在用的 (session > user-default > fallback)
        const current = getCurrentProfile()
        if (!current) {
          emitJson({ name: null, source: null, profile: null })
          return { handled: true, exitCode: 0 }
        }
        emitJson({ name: current.name, source: current.source, profile: desensitizeProfile(current.profile) })
        return { handled: true, exitCode: 0 }
      }
    }

    const scope = parseScope(args)

    // --set-model-profile <name>  (切换 active)
    {
      const idx = findFlag(args, '--set-model-profile')
      if (idx !== -1) {
        const name = args[idx + 1]
        if (!name || name.startsWith('--')) {
          emitError('--set-model-profile requires a <name> argument')
          return { handled: true, exitCode: 1 }
        }
        try {
          const r = setActiveProfile(name, scope)
          emitJson({
            ok: true,
            activeProfile: r.activeProfile,
            source: r.source,
            profile: desensitizeProfile(r.profile),
            scope,
          })
          return { handled: true, exitCode: 0 }
        } catch (e) {
          emitError((e as Error).message)
          return { handled: true, exitCode: 1 }
        }
      }
    }

    // --add-model-profile <name> --provider X --baseURL Y --model M --apiKey K [--name N]
    {
      const idx = findFlag(args, '--add-model-profile')
      if (idx !== -1) {
        const name = args[idx + 1]
        if (!name || name.startsWith('--')) {
          emitError('--add-model-profile requires a <name> argument')
          return { handled: true, exitCode: 1 }
        }
        if (getProfileByName(name)) {
          emitError(`profile "${name}" already exists; use --update-model-profile to modify`)
          return { handled: true, exitCode: 1 }
        }
        const provider = getOptionValue(args, '--provider')
        const baseURL = getOptionValue(args, '--baseURL')
        const model = getOptionValue(args, '--model')
        const apiKey = getOptionValue(args, '--apiKey')
        const displayName = getOptionValue(args, '--name')

        const missing: string[] = []
        if (!provider) missing.push('--provider')
        if (!baseURL) missing.push('--baseURL')
        if (!model) missing.push('--model')
        if (!apiKey) missing.push('--apiKey')
        if (missing.length) {
          emitError(`--add-model-profile missing required: ${missing.join(', ')}`)
          return { handled: true, exitCode: 1 }
        }
        if (!(PROFILE_PROVIDER_VALUES as readonly string[]).includes(provider!)) {
          emitError(`--provider must be one of ${PROFILE_PROVIDER_VALUES.join('|')}, got "${provider}"`)
          return { handled: true, exitCode: 1 }
        }
        const schema: ProfileSchema = {
          provider: provider as ProfileProvider,
          baseURL: baseURL!,
          model: model!,
          apiKey: apiKey!,
          ...(displayName ? { name: displayName } : {}),
        }
        try {
          setProfile(name, schema, scope)
          emitJson({
            ok: true,
            action: 'add',
            name,
            profile: desensitizeProfile(schema),
            scope,
          })
          return { handled: true, exitCode: 0 }
        } catch (e) {
          emitError((e as Error).message)
          return { handled: true, exitCode: 1 }
        }
      }
    }

    // --update-model-profile <name> [--baseURL Y] [--model M] [--apiKey K] [--name N]
    {
      const idx = findFlag(args, '--update-model-profile')
      if (idx !== -1) {
        const name = args[idx + 1]
        if (!name || name.startsWith('--')) {
          emitError('--update-model-profile requires a <name> argument')
          return { handled: true, exitCode: 1 }
        }
        const existing = getProfileByName(name)
        if (!existing) {
          emitError(`profile "${name}" not found; use --add-model-profile to create`)
          return { handled: true, exitCode: 1 }
        }
        const baseURL = getOptionValue(args, '--baseURL') ?? existing.baseURL
        const model = getOptionValue(args, '--model') ?? existing.model
        const apiKey = getOptionValue(args, '--apiKey') ?? existing.apiKey
        const displayName = getOptionValue(args, '--name') ?? existing.name
        const provider = getOptionValue(args, '--provider') ?? existing.provider
        const schema: ProfileSchema = {
          provider: provider as ProfileProvider,
          baseURL,
          model,
          apiKey,
          ...(displayName ? { name: displayName } : {}),
        }
        try {
          setProfile(name, schema, scope)
          emitJson({
            ok: true,
            action: 'update',
            name,
            profile: desensitizeProfile(schema),
            scope,
          })
          return { handled: true, exitCode: 0 }
        } catch (e) {
          emitError((e as Error).message)
          return { handled: true, exitCode: 1 }
        }
      }
    }

    // --set-model-profile-key <name> <key>
    {
      const idx = findFlag(args, '--set-model-profile-key')
      if (idx !== -1) {
        const name = args[idx + 1]
        const key = args[idx + 2]
        if (!name || name.startsWith('--') || !key || key.startsWith('--')) {
          emitError('--set-model-profile-key requires <name> <key> arguments')
          return { handled: true, exitCode: 1 }
        }
        const existing = getProfileByName(name)
        if (!existing) {
          emitError(`profile "${name}" not found`)
          return { handled: true, exitCode: 1 }
        }
        try {
          const updated: ProfileSchema = { ...existing, apiKey: key }
          setProfile(name, updated, scope)
          emitJson({
            ok: true,
            action: 'set-key',
            name,
            profile: desensitizeProfile(updated),
            scope,
          })
          return { handled: true, exitCode: 0 }
        } catch (e) {
          emitError((e as Error).message)
          return { handled: true, exitCode: 1 }
        }
      }
    }

    // --delete-model-profile <name>
    {
      const idx = findFlag(args, '--delete-model-profile')
      if (idx !== -1) {
        const name = args[idx + 1]
        if (!name || name.startsWith('--')) {
          emitError('--delete-model-profile requires a <name> argument')
          return { handled: true, exitCode: 1 }
        }
        const r = deleteProfile(name, scope)
        emitJson({
          ok: true,
          action: 'delete',
          name,
          deleted: r.deleted,
          activeProfileCleared: r.activeProfileCleared,
          remainingProfiles: Object.keys(r.profiles).sort(),
          scope,
        })
        return { handled: true, exitCode: r.deleted ? 0 : 1 }
      }
    }

    // --migrate-fallback-profile [--scope user|project] [--name <override>] [--force]
    //                            [--activate auto|always|never]
    // S1-09 收口: 把 env fallback 升级为正式 settings profile.
    {
      const idx = findFlag(args, '--migrate-fallback-profile')
      if (idx !== -1) {
        const targetName = getOptionValue(args, '--name')
        const force = args.includes('--force')
        const activateRaw = getOptionValue(args, '--activate')
        let activate: 'auto' | 'always' | 'never' = 'auto'
        if (activateRaw === 'always' || activateRaw === 'never' || activateRaw === 'auto') {
          activate = activateRaw
        } else if (activateRaw !== undefined) {
          emitError(`--activate must be one of auto|always|never, got "${activateRaw}"`)
          return { handled: true, exitCode: 1 }
        }
        const result = migrateFallbackProfile({
          scope,
          ...(targetName ? { targetName } : {}),
          force,
          activate,
        })
        if (result.ok === false) {
          emitError(result.reason)
          return { handled: true, exitCode: 1 }
        }
        if (result.migrated === true) {
          const written = getProfileByName(result.profileName)
          emitJson({
            ok: true,
            action: 'migrate',
            migrated: true,
            profileName: result.profileName,
            activeProfileSet: result.activeProfileSet,
            scope: result.scope,
            profile: written ? desensitizeProfile(written) : null,
          })
          return { handled: true, exitCode: 0 }
        }
        emitJson({
          ok: true,
          action: 'migrate',
          migrated: false,
          reason: result.reason,
          ...(result.profileName ? { profileName: result.profileName } : {}),
          scope: result.scope,
        })
        return { handled: true, exitCode: 0 }
      }
    }

    // --test-model-profile <name> [--timeout <ms>]
    // 真发 GET <baseURL>/models 验连通性 + apiKey 透传 (Workbench 测试连接).
    {
      const idx = findFlag(args, '--test-model-profile')
      if (idx !== -1) {
        const name = args[idx + 1]
        if (!name || name.startsWith('--')) {
          emitError('--test-model-profile requires a <name> argument')
          return { handled: true, exitCode: 1 }
        }
        const profile = getProfileByName(name)
        if (!profile) {
          emitError(`profile "${name}" not found`)
          return { handled: true, exitCode: 1 }
        }
        const timeoutStr = getOptionValue(args, '--timeout')
        const timeoutMs = timeoutStr ? Number.parseInt(timeoutStr, 10) : undefined
        const result = await testProfile(profile, timeoutMs ? { timeoutMs } : undefined)
        emitJson({
          ok: result.ok,
          action: 'test',
          name,
          profile: desensitizeProfile(profile),
          result,
        })
        return { handled: true, exitCode: result.ok ? 0 : 1 }
      }
    }

    return { handled: false, exitCode: 0 }
  } catch (e) {
    emitError(`internal error: ${(e as Error).message}`)
    return { handled: true, exitCode: 2 }
  }
}

File: crates/mossen-agent/src/services/config/profile_cli.rs
//! Profile CLI flag handler — handles --list-model-profiles, --get-model-profile, etc.

use serde_json::{json, Value};

use super::facade::{resolve_mossen_config, set_mossen_config_override};
use super::profiles::{
    delete_profile, desensitize_profile, desensitize_profiles, get_active_profile_name,
    get_current_profile, get_default_profile, get_fallback_profile, get_profile_by_name,
    get_profiles, list_all_profiles, migrate_fallback_profile, set_active_profile, set_profile,
    test_profile, validate_profile_name, ProfileProvider, ProfileSchema, ProfileSource,
    PROFILE_PROVIDER_VALUES,
};
use super::types::ConfigOverrideScope;

/// Check if any model profile CLI flag is present.
pub fn is_model_profile_flag_present(args: &[String]) -> bool {
    const FLAGS: &[&str] = &[
        "--list-model-profiles",
        "--get-model-profile",
        "--set-model-profile",
        "--add-model-profile",
        "--update-model-profile",
        "--set-model-profile-key",
        "--delete-model-profile",
        "--test-model-profile",
        "--migrate-fallback-profile",
    ];
    args.iter().any(|arg| FLAGS.contains(&arg.as_str()))
}

fn find_flag(args: &[String], flag: &str) -> Option<usize> {
    args.iter().position(|a| a == flag)
}

fn get_option_value<'a>(args: &'a [String], flag: &str) -> Option<&'a str> {
    let idx = args.iter().position(|a| a == flag)?;
    let next = args.get(idx + 1)?;
    if next.starts_with("--") {
        return None;
    }
    Some(next.as_str())
}

fn parse_scope(args: &[String]) -> ConfigOverrideScope {
    match get_option_value(args, "--scope") {
        Some("project") => ConfigOverrideScope::Project,
        _ => ConfigOverrideScope::User,
    }
}

/// Handle model profile CLI flags. Returns (handled, exit_code).
pub async fn handle_model_profile_cli_flag(args: &[String]) -> (bool, i32) {
    if !is_model_profile_flag_present(args) {
        return (false, 0);
    }

    // --list-model-profiles
    if find_flag(args, "--list-model-profiles").is_some() {
        let settings_profiles = get_profiles();
        let desensitized_settings = desensitize_profiles(&settings_profiles);
        let all = list_all_profiles();
        let fallback = get_fallback_profile();
        let current = get_current_profile();
        let default_p = get_default_profile();
        let all_with_source: Vec<Value> = all
            .iter()
            .map(|item| {
                json!({
                    "name": item.name,
                    "source": item.source,
                    "profile": desensitize_profile(&item.profile),
                })
            })
            .collect();
        let output = json!({
            "profiles": desensitized_settings,
            "activeProfile": get_active_profile_name(),
            "allProfiles": all_with_source,
            "fallbackProfile": fallback.as_ref().map(|fb| json!({
                "name": fb.name,
                "source": fb.source,
                "profile": desensitize_profile(&fb.profile),
            })),
            "currentProfile": current.as_ref().map(|c| json!({
                "name": c.name,
                "source": c.source,
                "profile": desensitize_profile(&c.profile),
            })),
            "defaultProfile": default_p.as_ref().map(|d| json!({
                "name": d.name,
                "source": d.source,
                "profile": desensitize_profile(&d.profile),
            })),
            "count": desensitized_settings.len(),
            "countAll": all_with_source.len(),
        });
        println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
        return (true, 0);
    }

    // --get-model-profile [<name>]
    if let Some(idx) = find_flag(args, "--get-model-profile") {
        let explicit_name = args.get(idx + 1).filter(|s| !s.starts_with("--"));
        if let Some(target_name) = explicit_name {
            if let Some(p) = get_profile_by_name(target_name) {
                let output = json!({"name": target_name, "source": "settings", "profile": desensitize_profile(&p)});
                println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
                return (true, 0);
            }
            if let Some(fb) = get_fallback_profile() {
                if fb.name == *target_name {
                    let output = json!({"name": target_name, "source": fb.source, "profile": desensitize_profile(&fb.profile)});
                    println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
                    return (true, 0);
                }
            }
            eprintln!("error: profile \"{}\" not found", target_name);
            return (true, 1);
        }
        let current = get_current_profile();
        match current {
            None => {
                println!("{}", json!({"name": null, "source": null, "profile": null}));
            }
            Some(c) => {
                let output = json!({"name": c.name, "source": c.source, "profile": desensitize_profile(&c.profile)});
                println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
            }
        }
        return (true, 0);
    }

    let scope = parse_scope(args);

    // --set-model-profile <name>
    if let Some(idx) = find_flag(args, "--set-model-profile") {
        let name = args.get(idx + 1).filter(|s| !s.starts_with("--"));
        match name {
            None => {
                eprintln!("error: --set-model-profile requires a <name> argument");
                return (true, 1);
            }
            Some(name) => match set_active_profile(name, scope) {
                Ok((active, profile, source)) => {
                    let output = json!({
                        "ok": true,
                        "activeProfile": active,
                        "source": source,
                        "profile": desensitize_profile(&profile),
                        "scope": format!("{:?}", scope).to_lowercase(),
                    });
                    println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
                    return (true, 0);
                }
                Err(e) => {
                    eprintln!("error: {}", e);
                    return (true, 1);
                }
            },
        }
    }

    // --add-model-profile <name>
    if let Some(idx) = find_flag(args, "--add-model-profile") {
        let name = args.get(idx + 1).filter(|s| !s.starts_with("--"));
        match name {
            None => {
                eprintln!("error: --add-model-profile requires a <name> argument");
                return (true, 1);
            }
            Some(name) => {
                if get_profile_by_name(name).is_some() {
                    eprintln!("error: profile \"{}\" already exists; use --update-model-profile to modify", name);
                    return (true, 1);
                }
                let provider = get_option_value(args, "--provider");
                let base_url = get_option_value(args, "--baseURL");
                let model = get_option_value(args, "--model");
                let api_key = get_option_value(args, "--apiKey");
                let display_name = get_option_value(args, "--name");

                let mut missing = Vec::new();
                if provider.is_none() { missing.push("--provider"); }
                if base_url.is_none() { missing.push("--baseURL"); }
                if model.is_none() { missing.push("--model"); }
                if api_key.is_none() { missing.push("--apiKey"); }
                if !missing.is_empty() {
                    eprintln!("error: --add-model-profile missing required: {}", missing.join(", "));
                    return (true, 1);
                }

                let provider = provider.unwrap();
                if !PROFILE_PROVIDER_VALUES.contains(&provider) {
                    eprintln!("error: --provider must be one of {}, got \"{}\"", PROFILE_PROVIDER_VALUES.join("|"), provider);
                    return (true, 1);
                }

                let schema = json!({
                    "provider": provider,
                    "baseURL": base_url.unwrap(),
                    "model": model.unwrap(),
                    "apiKey": api_key.unwrap(),
                    "name": display_name,
                });

                match set_profile(name, &schema, scope) {
                    Ok(_) => {
                        let output = json!({"ok": true, "action": "add", "name": name});
                        println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
                        return (true, 0);
                    }
                    Err(e) => {
                        eprintln!("error: {}", e);
                        return (true, 1);
                    }
                }
            }
        }
    }

    // --update-model-profile <name>
    if let Some(idx) = find_flag(args, "--update-model-profile") {
        let name = args.get(idx + 1).filter(|s| !s.starts_with("--"));
        match name {
            None => {
                eprintln!("error: --update-model-profile requires a <name> argument");
                return (true, 1);
            }
            Some(name) => {
                let existing = match get_profile_by_name(name) {
                    Some(p) => p,
                    None => {
                        eprintln!("error: profile \"{}\" not found; use --add-model-profile to create", name);
                        return (true, 1);
                    }
                };
                let base_url = get_option_value(args, "--baseURL").unwrap_or(&existing.base_url);
                let model = get_option_value(args, "--model").unwrap_or(&existing.model);
                let api_key = get_option_value(args, "--apiKey").unwrap_or(&existing.api_key);
                let display_name = get_option_value(args, "--name").or(existing.name.as_deref());
                let provider_str = get_option_value(args, "--provider").unwrap_or("openai-compatible");

                let schema = json!({
                    "provider": provider_str,
                    "baseURL": base_url,
                    "model": model,
                    "apiKey": api_key,
                    "name": display_name,
                });

                match set_profile(name, &schema, scope) {
                    Ok(_) => {
                        let output = json!({"ok": true, "action": "update", "name": name});
                        println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
                        return (true, 0);
                    }
                    Err(e) => {
                        eprintln!("error: {}", e);
                        return (true, 1);
                    }
                }
            }
        }
    }

    // --set-model-profile-key <name> <key>
    if let Some(idx) = find_flag(args, "--set-model-profile-key") {
        let name = args.get(idx + 1).filter(|s| !s.starts_with("--"));
        let key = args.get(idx + 2).filter(|s| !s.starts_with("--"));
        match (name, key) {
            (Some(name), Some(key)) => {
                let existing = match get_profile_by_name(name) {
                    Some(p) => p,
                    None => {
                        eprintln!("error: profile \"{}\" not found", name);
                        return (true, 1);
                    }
                };
                let updated = json!({
                    "provider": "openai-compatible",
                    "baseURL": existing.base_url,
                    "model": existing.model,
                    "apiKey": key,
                    "name": existing.name,
                });
                match set_profile(name, &updated, scope) {
                    Ok(_) => {
                        let output = json!({"ok": true, "action": "set-key", "name": name});
                        println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
                        return (true, 0);
                    }
                    Err(e) => {
                        eprintln!("error: {}", e);
                        return (true, 1);
                    }
                }
            }
            _ => {
                eprintln!("error: --set-model-profile-key requires <name> <key> arguments");
                return (true, 1);
            }
        }
    }

    // --delete-model-profile <name>
    if let Some(idx) = find_flag(args, "--delete-model-profile") {
        let name = args.get(idx + 1).filter(|s| !s.starts_with("--"));
        match name {
            None => {
                eprintln!("error: --delete-model-profile requires a <name> argument");
                return (true, 1);
            }
            Some(name) => {
                let (deleted, active_cleared, remaining) = delete_profile(name, scope);
                let remaining_names: Vec<&String> = remaining.keys().collect();
                let output = json!({
                    "ok": true,
                    "action": "delete",
                    "name": name,
                    "deleted": deleted,
                    "activeProfileCleared": active_cleared,
                    "remainingProfiles": remaining_names,
                });
                println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
                return (true, if deleted { 0 } else { 1 });
            }
        }
    }

    // --migrate-fallback-profile
    if find_flag(args, "--migrate-fallback-profile").is_some() {
        let target_name = get_option_value(args, "--name");
        let force = args.iter().any(|a| a == "--force");
        let activate_raw = get_option_value(args, "--activate");
        let activate = match activate_raw {
            Some("always") => "always",
            Some("never") => "never",
            Some("auto") | None => "auto",
            Some(other) => {
                eprintln!("error: --activate must be one of auto|always|never, got \"{}\"", other);
                return (true, 1);
            }
        };
        let result = migrate_fallback_profile(scope, target_name, force, activate);
        let output = serde_json::to_value(&result).unwrap_or(Value::Null);
        println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
        return (true, 0);
    }

    // --test-model-profile <name>
    if let Some(idx) = find_flag(args, "--test-model-profile") {
        let name = args.get(idx + 1).filter(|s| !s.starts_with("--"));
        match name {
            None => {
                eprintln!("error: --test-model-profile requires a <name> argument");
                return (true, 1);
            }
            Some(name) => {
                let profile = match get_profile_by_name(name) {
                    Some(p) => p,
                    None => {
                        eprintln!("error: profile \"{}\" not found", name);
                        return (true, 1);
                    }
                };
                let timeout_str = get_option_value(args, "--timeout");
                let timeout_ms = timeout_str.and_then(|s| s.parse::<u64>().ok());
                let result = test_profile(&profile, timeout_ms).await;
                let output = json!({
                    "ok": result.ok,
                    "action": "test",
                    "name": name,
                    "profile": desensitize_profile(&profile),
                    "result": result,
                });
                println!("{}", serde_json::to_string_pretty(&output).unwrap_or_default());
                return (true, if result.ok { 0 } else { 1 });
            }
        }
    }

    (false, 0)
}

/// Handle config CLI flags (--get/set/clear/list-mossen-config).
pub async fn handle_config_cli_flag(args: &[String]) -> (bool, i32) {
    let get_idx = args.iter().position(|a| a == "--get-mossen-config");
    let set_idx = args.iter().position(|a| a == "--set-mossen-config");
    let clear_idx = args.iter().position(|a| a == "--clear-mossen-config");
    let list_idx = args.iter().position(|a| a == "--list-mossen-config");

    if get_idx.is_none() && set_idx.is_none() && clear_idx.is_none() && list_idx.is_none() {
        return (false, 0);
    }

    let scope_idx = args.iter().position(|a| a == "--scope");
    let scope_str = scope_idx
        .and_then(|i| args.get(i + 1))
        .map(|s| s.as_str())
        .unwrap_or("override");
    let scope = match scope_str {
        "user" => ConfigOverrideScope::User,
        "project" => ConfigOverrideScope::Project,
        "override" => ConfigOverrideScope::Override,
        other => {
            eprintln!("error: invalid --scope \"{}\"; expected user|project|override", other);
            return (true, 1);
        }
    };

    if let Some(idx) = get_idx {
        let key = match args.get(idx + 1) {
            Some(k) => k,
            None => {
                eprintln!("error: --get-mossen-config requires a key argument");
                return (true, 1);
            }
        };
        let r = resolve_mossen_config(key, Value::Null);
        println!("{}", serde_json::to_string(&r.value).unwrap_or_default());
        return (true, 0);
    }

    if let Some(idx) = set_idx {
        let key = args.get(idx + 1);
        let value_str = args.get(idx + 2);
        match (key, value_str) {
            (Some(k), Some(v)) => {
                let parsed: Value = match serde_json::from_str(v) {
                    Ok(p) => p,
                    Err(e) => {
                        eprintln!("error: --set-mossen-config value must be valid JSON: {}", e);
                        return (true, 1);
                    }
                };
                set_mossen_config_override(k, parsed, scope);
                return (true, 0);
            }
            _ => {
                eprintln!("error: --set-mossen-config requires <key> <value-as-json>");
                return (true, 1);
            }
        }
    }

    if let Some(idx) = clear_idx {
        let key = match args.get(idx + 1) {
            Some(k) => k,
            None => {
                eprintln!("error: --clear-mossen-config requires a key argument");
                return (true, 1);
            }
        };
        super::facade::clear_mossen_config_overrides(scope, Some(key));
        return (true, 0);
    }

    if list_idx.is_some() {
        let all = super::facade::get_all_mossen_config_values();
        println!("{}", serde_json::to_string_pretty(&all).unwrap_or_default());
        return (true, 0);
    }

    (false, 0)
}

/// TS `type ModelProfileFlag` — variants emitted by the CLI flag parser.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ModelProfileFlag {
    /// `--model-profile <name>`
    Named(String),
    /// `--model-profile=current` reuses the active profile.
    UseCurrent,
}
